contrib/kae/private_key_providers/source/kae_private_key_provider.cc

```diff
@@ -0,0 +1,399 @@
+#include "source/common/config/datasource.h"
+
+#include "contrib/kae/private_key_providers/source/kae.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PrivateKeyMethodProvider {
+namespace Kae {
+
+SINGLETON_MANAGER_REGISTRATION(kae_manager);
+
+// KaePrivateKeyConnection
+KaePrivateKeyConnection::KaePrivateKeyConnection(Ssl::PrivateKeyConnectionCallbacks& cb,
+                                                 Event::Dispatcher& dispatcher, KaeHandle& handle,
+                                                 bssl::UniquePtr<EVP_PKEY> pkey)
+    : cb_(cb), dispatcher_(dispatcher), handle_(handle), pkey_(std::move(pkey)) {}
+
+void KaePrivateKeyConnection::registerCallback(KaeContext* ctx) {
+
+  // Get the receiving end of the notification pipe. The other end is written to by the polling
+  // thread.
+  int fd = ctx->getFd();
+
+  ssl_async_event_ = dispatcher_.createFileEvent(
+      fd,
+      [this, ctx, fd](uint32_t) {
+        int status = WD_STATUS_FAILED;
+        {
+          Thread::LockGuard data_lock(ctx->data_lock_);
+          int bytes = read(fd, &status, sizeof(status));
+          if (bytes != sizeof(status)) {
+            status = WD_STATUS_FAILED;
+          }
+          if (status == WD_STATUS_BUSY) {
+            // At this point we are no longer allowed to have the status as "busy", because the
+            // upper levels consider the operation complete.
+            status = WD_STATUS_FAILED;
+          }
+          ctx->setOpStatus(status);
+        }
+        this->cb_.onPrivateKeyMethodComplete();
+        return absl::OkStatus();
+      },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+}
+
+void KaePrivateKeyConnection::unregisterCallback() { ssl_async_event_ = nullptr; }
+
+namespace {
+ssl_private_key_result_t privateKeySignInternal(SSL* ssl, KaePrivateKeyConnection* ops, uint8_t*,
+                                                size_t*, size_t, uint16_t signature_algorithm,
+                                                const uint8_t* in, size_t in_len) {
+  RSA* rsa;
+  const EVP_MD* md;
+  bssl::ScopedEVP_MD_CTX ctx;
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int hash_len;
+  uint8_t* msg;
+  size_t msg_len;
+  int prefix_allocated = 0;
+  KaeContext* kae_ctx = nullptr;
+  int padding = RSA_NO_PADDING;
+
+  if (ops == nullptr) {
+    return ssl_private_key_failure;
+  }
+
+  KaeHandle& kae_handle = ops->getHandle();
+  EVP_PKEY* rsa_pkey = ops->getPrivateKey();
+
+  if (rsa_pkey == nullptr) {
+    goto error;
+  }
+  if (EVP_PKEY_id(rsa_pkey) != SSL_get_signature_algorithm_key_type(signature_algorithm)) {
+    goto error;
+  }
+
+  rsa = EVP_PKEY_get0_RSA(rsa_pkey);
+  if (rsa == nullptr) {
+    goto error;
+  }
+  md = SSL_get_signature_algorithm_digest(signature_algorithm);
+  if (md == nullptr) {
+    goto error;
+  }
+
+  // Create KAE context which will be used for this particular signing/decryption.
+  kae_ctx = new KaeContext(kae_handle);
```

Comment:
Please prefer smart pointers like `std::unique_ptr` or `std::shared_ptr` over raw `new` and `delete`.